firebase-sessions/src/main/kotlin/com/google/firebase/sessions/leader/SessionMaintainerLeader.kt

```diff
@@ -0,0 +1,170 @@
+  private val sessionGenerator =
+    SessionGenerator(collectEvents = shouldCollectEvents(), timeProvider)
+  private val eventGDTLogger = EventGDTLogger(transportFactoryProvider)
+  private val sessionCoordinator = SessionCoordinator(firebaseInstallations, eventGDTLogger)
+
+  private val TAG = "SessionMaintainerLeader"
+
+  override fun start(backgroundDispatcher: CoroutineDispatcher) {
+    val sessionInitiateListener =
+      object : SessionInitiateListener {
+        // Avoid making a public function in FirebaseSessions for onInitiateSession.
+        override suspend fun onInitiateSession(sessionDetails: SessionDetails) {
+          initiateSessionStart(sessionDetails)
+        }
+      }
+
+    val sessionInitiator =
+      SessionInitiator(
+        timeProvider,
+        backgroundDispatcher,
+        sessionInitiateListener,
+        sessionSettings,
+        sessionGenerator,
+      )
+
+    val appContext = firebaseApp.applicationContext.applicationContext
+    if (appContext is Application) {
+      appContext.registerActivityLifecycleCallbacks(sessionInitiator.activityLifecycleCallbacks)
+
+      firebaseApp.addLifecycleEventListener { _, _ ->
+        Log.w(TAG, "FirebaseApp instance deleted. Sessions library will not collect session data.")
+        appContext.unregisterActivityLifecycleCallbacks(sessionInitiator.activityLifecycleCallbacks)
+      }
+    } else {
+      Log.e(
+        TAG,
+        "Failed to register lifecycle callbacks, unexpected context ${appContext.javaClass}."
+      )
+    }
+  }
+  override fun register(subscriber: SessionSubscriber) {
+    /*
+     * Immediately call the callback if Sessions generated a session before the
+     * subscriber subscribed, otherwise subscribers might miss the first session.
+     */
+    if (sessionGenerator.hasGenerateSession) {
+      subscriber.onSessionChanged(
+        SessionSubscriber.SessionDetails(sessionGenerator.currentSession.sessionId)
+      )
+    }
+  }
+
+  private suspend fun initiateSessionStart(sessionDetails: SessionDetails) {
+    val subscribers = FirebaseSessionsDependencies.getRegisteredSubscribers()
+
+    if (subscribers.isEmpty()) {
+      Log.d(
+        TAG,
+        "Sessions SDK did not have any dependent SDKs register as dependencies. Events will not be sent."
+      )
+      return
+    }
+
+    subscribers.values.forEach { subscriber ->
+      // Notify subscribers, regardless of sampling and data collection state.
+      subscriber.onSessionChanged(SessionSubscriber.SessionDetails(sessionDetails.sessionId))
+    }
+
+    if (subscribers.values.none { it.isDataCollectionEnabled }) {
+      Log.d(TAG, "Data Collection is disabled for all subscribers. Skipping this Session Event")
+      return
+    }
+
+    Log.d(TAG, "Data Collection is enabled for at least one Subscriber")
+
+    // This will cause remote settings to be fetched if the cache is expired.
+    sessionSettings.updateSettings()
```

Comment:
I assume this is asynchronous. So anything below will execute based on the previously cached value of settings fetch?

Reply:
No.

Reply:
In that case, is the updateSettings a blocking call? Was curious on the scenario of when will a settings change be reflecting on the client from the moment of change on the backend.

Reply:
It's a suspend function and it'll pause until the settings are fetched. Then after this line, settings have been updated.